offer-manager-app/src/main/groovy/life/qbic/portal/offermanager/offergeneration/OfferTemplate.groovy

```diff
@@ -0,0 +1,370 @@
+        document.getElementById("project-title").text(offer.getProjectTitle())
+        document.getElementById("project-description").text(offer.getProjectObjective())
+        if (offer.getExperimentalDesign()) {
+            document.getElementById("experimental-design").text(offer.getExperimentalDesign())
+        }
+    }
+
+    private static void fillCostSummary(Document document, OfferContent offer) {
+        // the overhead costs
+        document.select(".overheads  .data-generation  > .costs").each {element ->
+            element.text(Currency.format(offer.getOverheadsDataGeneration()))
+        }
+        document.select(".overheads  .data-analysis  > .costs").each {element ->
+            element.text(Currency.format(offer.getOverheadsDataAnalysis()))
+        }
+        document.select(".overheads  .data-management  > .costs").each {element ->
+            element.text(Currency.format(offer.getOverheadsProjectManagementAndDataStorage()))
+        }
+        document.select(".overheads  > .totals-section  > .costs").each {element ->
+            element.text(Currency.format(offer.getOverheadTotal()))
+        }
+        // other fields
+        document.select(".total-costs .net > .costs").each {element ->
+            element.text(Currency.format(offer.getNetCost()))
+        }
+        document.select(".total-costs .discounts > .costs").each {element ->
+            element.text(Currency.format(offer.getTotalDiscountAmount() * (-1)))
+        }
+        document.select(".total-costs .overheads > .costs").each {element ->
+            element.text(Currency.format(offer.getOverheadTotal()))
+        }
+        document.select(".total-costs .vat > .costs").each {element ->
+            element.text(Currency.format(offer.getTotalVat()))
+        }
+        document.select(".total-costs > .totals-section > .costs").each {element ->
+            element.text(Currency.format(offer.getTotalCost()))
+        }
+    }
+
+    private static void fillDataGenerationItems(Document document, OfferContent offer) {
+        Element dataGenerationTable = document.getElementById("data-generation-items")
+
+        // remove table if no items are available
+        if (offer.getDataGenerationItems().isEmpty()) {
+            dataGenerationTable.remove()
+            return
+        }
+        // clear the table
+        dataGenerationTable.select("tbody > tr").each {element ->
+            element.remove()
+        }
+        // fill the table
+        int position = 1
+        Element tableBody = dataGenerationTable.selectFirst("tbody")
+        for (OfferItem offerItem in offer.getDataGenerationItems()) {
+            OfferPosition offerPosition = OfferPosition.createProductItem(position, offerItem)
+            tableBody.append(offerPosition.outerHtml())
+            position++
+            if (offerItem.getQuantityDiscount() > 0) {
+                OfferPosition discountPosition = OfferPosition.createDiscount(position, position - 1, offerItem)
```

Comment:
See my comment above. Furthermore, do you really need to provide position -1. Since we agreed on displaying the discount after the item always I think this can be also achieved with only providing the position itself

Reply:
It is implemented this way because the logic on where the discount is listed is decided by the template, not the row itself. The OfferPosition is only responsible for providing the HTML for a discount for a selected position on the offer.

Reply:
ok.. but then, instead of providing two positions, which is really confusing to read, how about providing the discount descriptions as string? This is why you need the second position, right? And then the offer template determines how the text looks like.

Reply:
Sadly this does not work. The description has to span multiple columns so we need two rows here.

Reply:
You can look at it this way: The OfferItem is the DTO and the OfferPosition the logical entity